File: medium/0300_longest_increasing_subsequence/solution.py
from bisect import bisect_left


class Solution:
    def lengthOfLIS(self, nums: list[int]) -> int:
        tails = [nums[0]]

        for num in nums:
            if num > tails[-1]:
                tails.append(num)
            elif num < tails[-1]:
                tails[bisect_left(tails, num)] = num

        return len(tails)

File: medium/0300_longest_increasing_subsequence/solution.cpp
#include <vector>

class Solution {
public:
  int lengthOfLIS(std::vector<int> &nums) {
    std::vector<int> dp(nums.size(), 1);

    for (size_t i{}; i < nums.size(); ++i) {
      for (size_t j = i - 1; j != -1; --j) {
        if (nums[j] < nums[i])
          dp[i] = std::max(dp[i], dp[j] + 1);
      }
    }

    return *max_element(dp.begin(), dp.end());
  }
};
